tests/django_pg_migration_tools/test_operations.py

```diff
@@ -427,3 +454,355 @@ def test_when_not_allowed_to_migrate(self):
+                operation.database_forwards(
+                    self.app_label, editor, project_state, new_state
+                )
+
+        # No queries have run, because the migration wasn't allowed to run by
+        # the router.
+        assert len(queries) == 0
+
+        # Try the same for the reverse operation:
+        # Proceed to try and add the index + constraint:
+        with connection.schema_editor(atomic=False, collect_sql=False) as editor:
+            with utils.CaptureQueriesContext(connection) as queries:
+                operation.database_backwards(
+                    self.app_label, editor, project_state, new_state
+                )
+
+        # No queries have run, because the migration wasn't allowed to run by
+        # the router.
+        assert len(queries) == 0
+
+    @pytest.mark.django_db(transaction=True)
+    def test_raises_if_constraint_already_exists(self):
+        project_state = ProjectState()
+        project_state.add_model(ModelState.from_model(IntModel))
+        new_state = project_state.clone()
+
+        # Create the constraint so that the operation raises when we try to
+        # recreate the constraint with the raise_if_exists flag set to True.
+        with connection.cursor() as cursor:
+            cursor.execute(_CREATE_CONSTRAINT_QUERY)
+
+        operation = operations.SaferAddUniqueConstraint(
+            raise_if_exists=True,
+            model_name="intmodel",
+            constraint=UniqueConstraint(
+                fields=("int_field",),
+                name="unique_int_field",
+            ),
+        )
+
+        with pytest.raises(operations.ConstraintAlreadyExists):
+            with connection.schema_editor(atomic=False, collect_sql=False) as editor:
+                operation.database_forwards(
+                    self.app_label, editor, project_state, new_state
+                )
+
+        # Drop the constraint. We aren't in a test with transaction, we have
+        # to clean up.
+        with connection.cursor() as cursor:
+            cursor.execute(_DROP_CONSTRAINT_QUERY)
+
+    @pytest.mark.django_db(transaction=True)
+    def test_do_nothing_when_asked_not_to_raise_when_constraint_exists(self):
+        project_state = ProjectState()
+        project_state.add_model(ModelState.from_model(IntModel))
+        new_state = project_state.clone()
+
+        # Create the constraint. The operation won't raise an error when the
+        # constraint already exists because `raise_if_exists` is False.
+        with connection.cursor() as cursor:
+            cursor.execute(_CREATE_CONSTRAINT_QUERY)
+
+        operation = operations.SaferAddUniqueConstraint(
+            raise_if_exists=False,
+            model_name="intmodel",
+            constraint=UniqueConstraint(
+                fields=("int_field",),
+                name="unique_int_field",
+            ),
+        )
+
+        with connection.schema_editor(atomic=False, collect_sql=False) as editor:
```

Comment:
❓ So in this scenario this is our way to make it "idempotent", by causing the migration to "pass" even if the constraint already exists?

Reply:
Correct.

As this migration can be retried or used as part of a larger migration file with multiple non-atomic operations (any in which could fail and require retries on their own ways), this is a way to make sure we don't blow up and require manual database intervention if the constraint already exists.